wrappers/c/zxing.cpp

```diff
@@ -0,0 +1,159 @@
+#include "zxing.h"
+
+#include <ZXing/ImageView.h>
+#include <ZXing/BarcodeFormat.h>
+#include <ZXing/DecodeHints.h>
+#include <ZXing/Result.h>
+#include <ZXing/ReadBarcode.h>
+
+extern "C" {
+
+/*
+ * ZXing/ImageView.h
+ */
+
+zxing_ImageView zxing_ImageView_new (const uint8_t * data, int width, int height, zxing_ImageFormat format, int rowStride, int pixStride)
+{
+	ZXing::ImageFormat cppformat = (ZXing::ImageFormat) format;
```

Comment:
Please use a `reinterpret_cast<ZXing::ImageFormat>(format)` here. Same for all the other c-style casts below.

Reply:
I think it's time now to discuss enums.

Using `reinterpret_cast` in `zxing_ReadBarcode()`/`zxing_ReadBarcodes()` works fine, but for the enums the compiler complains with errors such as these:

```
zxing.cpp: In function ‘zxing_ImageView* zxing_ImageView_new(const uint8_t*, int, int, zxing_ImageFormat, int, int)’:
zxing.cpp:17:40: error: invalid cast from type ‘zxing_ImageFormat’ to type ‘ZXing::ImageFormat’
   17 |         ZXing::ImageFormat cppformat = reinterpret_cast<ZXing::ImageFormat>(format);
      |                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
zxing.cpp: In function ‘zxing_ImageFormat zxing_ImageView_format(zxing_ImageView*)’:
zxing.cpp:48:16: error: invalid cast from type ‘ZXing::ImageFormat’ to type ‘zxing_ImageFormat’
   48 |         return reinterpret_cast<zxing_ImageFormat>(iv->format());
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

---

According to [this](https://stackoverflow.com/questions/20792121) question and answers on SO it's as I suspected: there's no way to "export" C++ enums as C enums.

The alternatives are basically (a) to use plain C enums alone, or (b) to define both C and C++ enums and use a (runtime) translation between the two.

(a) may be the easier alternative but is not backwards compatible. Plus C++ features (such as namespaces and "object-like" enums) can't be used (I think?).

(b) The latter is backwards compatible and requires no changes on the C++ library, but obviously incurs a performance penalty (though C->C++ translation can be minimised to an array indexing operation).

The accepted answer suggests a sort of mix between the two, where both C and C++ enums are defined, with the same values, but the C enum has a typical C prefix, while the C++ enum has no prefix (because of namespaces). The real definition of the enums can be saved in a file of its own so that it can be included either by the C or the C++ headers.

Reply:
The compiler didn't complain about `static_cast`.

Are C++ enum variants guaranteed to be unique sequential integers (unless otherwise explicitly specified)?

```cpp
enum foo {
  variant0, // 0
  variant1, // 1
  variant2, // 2
  // ...
}
```

Reply:
Yes they are. There might be differences in the width of the underlying integer type. But the static_cast should take care of that.

Reply:
Ok nice, this is solved then.